ext/crates/fp/src/limb.rs

```diff
@@ -11,6 +10,148 @@ pub(crate) struct LimbBitIndexPair {
     pub(crate) bit_index: usize,
 }
 
+/// A struct that defines a range of entries in a slice of limbs. We number limbs starting from some
+/// other limb that might not be in the range. This is because we must be able to index a struct
+/// using indices for another ambient struct. For example, accessing a `Slice` using coordinates
+/// from an ambient `FpVector`.
+#[derive(Debug, Hash, PartialEq, Eq, Clone, Copy)]
+pub struct LimbLength<const P: u32> {
+    /// The index of the first entry relative to some other base limb.
+    pub(crate) start: usize,
+
+    /// The index of the last entry relative to some other base limb.
+    pub(crate) end: usize,
+
+    /// The total number of limbs in the range.
+    ///
+    /// We store this value instead of computing it on the fly because benchmarks tend to show that
+    /// the tradeoff is beneficial in high dimensions (>1000). We might want to only enable this
+    /// when odd-primes is enabled, since computing this number is easier when `p == 2`, so the
+    /// tradeoff is potentially worse.
+    limbs: usize,
+}
+
+impl<const P: u32> LimbLength<P> {
+    pub const fn from_limbs(limbs: usize) -> Self {
+        let logical = entries_per_limb_const::<P>() * limbs;
+        Self {
+            start: 0,
+            end: logical,
+            limbs,
+        }
+    }
+
+    pub const fn from_logical(logical: usize) -> Self {
+        let limbs = number::<P>(logical);
+        Self {
+            start: 0,
+            end: logical,
+            limbs,
+        }
+    }
+
+    /// Returns a `LimbLength` describing a vector starting at entry `start` and ending at entry
+    /// `end`. Most methods assume that `self.start < entries_per_limb_const::<P>()`, so it advised
+    /// to call `apply_shift` before using the return value.
```

Comment:
What exactly are these "most methods"? It seems like e.g. `bit_offset` makes such an assumption. I think this should be in the documentation of the functions that make these assumptions rather than `from_start_end`.

Reply:
I was a bit lazy to check if the assumption was needed in each case, since it's only an internal struct, but it should be good now

Reply:
I think these kind of unspoken assumptions can be very dangerous, especially if other people modify the code

Reply:
I think I added the documentation in the right places now

Reply:
I think the "it [is] advised to call `apply_shift` before using the return value" can now be removed

Reply:
Done